skills_ml/algorithms/skill_feature_creator/__init__.py

```diff
@@ -1,81 +1,121 @@
-            if not (set(features) < set(self.all_features)):
-                not_supported = list(set(features) - set(self.all_features))
-                raise Exception("{} not supported!".format(", ".join(not_supported)))
-            else:
-                features_to_be_combined = features
+        features_to_be_combined = self.selected_features
+        feature_objects = [FeatureFactory._factory(feature, **self.params) for feature in features_to_be_combined]
+        if combine == "concat":
+            for doc in docs:
+                yield reduce(lambda x, y: x+y, map(lambda x: x.output(doc), feature_objects))
         else:
-            raise Exception(TypeError)
+            raise Exception("\"{}\" not supported!".format(combine))
+
+
+class FeatureFactory(object):
+    @staticmethod
+    def _factory(type, **kwargs):
+        if type == "StructuralFeature":
+            return StructuralFeature()
+
+        if type == "ContextualFeature":
+            return ContextualFeature()
+
+        if type == "EmbeddingFeature":
+            return EmbeddingFeature(**kwargs)
 
-        result = []
-        for feature in features_to_be_combined:
-            result.extend(FeatureCreator._factory(feature, self.doc).output())
-        return result
+        else:
+            raise ValueError('Bad feature type \"{}\"'.format(type))
 
 
-class StructuralFeature(FeatureCreator):
+class StructuralFeature(FeatureFactory):
     """ Sturctural features
     """
-    def output(self):
-        """ Output a feature vector. Now it's just a simple example.
+    def output(self, doc):
+        """ Output a feature vector. Need to be implemented! Now it's just a simple example.
         """
-        return ["StructuralFeature" + "_" + self.doc]
+        return ["StructuralFeature_" + doc]
 
 
-class ContextualFeature(FeatureCreator):
+class ContextualFeature(FeatureFactory):
     """ Contextual features
     """
-    def output(self):
-        """ Output a feature vector. Now it's just a simple example.
+    def output(self, doc):
+        """ Output a feature vector. Need to be implemented! Now it's just a simple example.
         """
-        return ["ContextualFeature" + "_" + self.doc]
+        return ["ContextualFeature_" + doc]
 
 
-class EmbeddingFeature(FeatureCreator):
+class EmbeddingFeature(FeatureFactory):
     """ Embedding features
     """
-    def output(self):
-        """ Output a feature vector. Now it's just a simple example.
-        """
-        return ["EmbeddingFeature" + "_" + self.doc]
+    def __init__(self, **kwargs):
+        super().__init__()
+        model_name = kwargs["embedding_model_name"]
+        path = kwargs["embedding_model_path"]
+        s3_conn = kwargs["s3_conn"]
+        self.embedding_model = Doc2Vectorizer(model_name, path, s3_conn)
 
+    def output(self, doc):
+        """ Output a feature vector.
+        """
+        if type(self.embedding_model.model).__name__ == "Doc2Vec":
```

Comment:
Should we make this duck-typed? In other words, call `infer_vector` if there is an `infer_vector` available to call? To protect against name changes, or even just support other hypothetical models that implement the `infer_vector` interface.

Reply:
Oh, I had another thought looking at the code of this and `Doc2Vectorizer`. Instead of reaching into `self.embedding_model`'s model object to call a method, should the `self.embedding_model` just hide the implementation by exposing a method of its own? There is already a `Doc2Vectorize.vectorize` that takes many documents. Maybe that should be `vectorize_documents`, that calls a `vectorize_document` that outputs a feature vector for one document. The latter method could be called from here.

Whether or not this is a good idea might depend on how you plan on extending this. I'm picturing other objects with a similar interface to Doc2Vectorizer being usable here in the future, but if you only forsee Doc2Vectorizers of different paths than this may be irrelevant.

Reply:
what if I keep the method `vectorize` the same, but change the output of embeddingFeature to be `list(list(self.embedding_model.vectorize([doc]))[0])`  to output a feature vector for one document?

Reply:
That works

Reply:
I think this if statement can go away now, or at the very least assert self.embedding_model's class instead of that of the underlying .model object.

Reply:
yeah agree. The original reason of the if statement is because we might want to expand the Doc2Vectorizer class to not limit to only gensim model object but also be compatible with other libraries like tensorflow or pytorch. yes you're right, just let Doc2Vectorizer worry about it.